certstore_linux.go

```diff
@@ -1,14 +1,324 @@
+func (i *nssIdentity) Delete() error {
+	var (
+		cert      = C.CERTCertificate(*i)
+		secstatus = C.PK11_DeleteTokenCertAndKey(&cert, nil)
+	)
+	if secstatus != C.SECSuccess {
+		return errors.New("failed to delete certificate & key pair")
+	}
+	return nil
+}
+
+// Close implements the Identity interface.
+func (i *nssIdentity) Close() {
+	cert := C.CERTCertificate(*i)
+	C.CERT_DestroyCertificate(&cert)
+}
+
+// Public implements the crypto.Signer interface.
+func (i *nssIdentity) Public() crypto.PublicKey {
+	var cert, _ = i.Certificate()
+	if cert == nil {
+		return nil
+	}
+	return cert.PublicKey
+}
+
+// Sign implements the crypto.Signer interface.
+func (i *nssIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash := opts.HashFunc()
+	if len(digest) != hash.Size() {
+		return nil, errors.New("bad digest for hash")
+	}
+	var (
+		cert = C.CERTCertificate(*i)
+		key  = C.PK11_FindKeyByAnyCert(&cert, nil)
+	)
+	if key == nil {
+		return nil, errors.New("cannot find private key")
+	}
+
+	digestC := C.SECITEM_AllocItem(nil, nil, C.uint(len(digest)))
+	if digestC == nil {
+		return nil, errors.New("failure to allocate memory for digest")
+	}
+	defer C.SECITEM_FreeItem(digestC, 1)
+
+	resultC := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	if resultC == nil {
+		return nil, errors.New("failure to allocate memory for signature")
+	}
+	defer C.SECITEM_FreeItem(resultC, 1)
+
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
+		var dsaSig C.SECItem
+		if C.DSAU_EncodeDerSigWithLen(&dsaSig, resultC, resultC.len) != C.SECSuccess {
+			return nil, errors.New("unable to generate dsa signature")
+		}
+		resultC = &dsaSig
+	}
+	return C.GoBytes(unsafe.Pointer(resultC.data), C.int(resultC.len)), nil
+}
+
+// Import implements the Store interface.
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	pass := C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("error allocating memory for PKCS#12 password")
+	}
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	p12 := C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+	if p12 == nil {
+		return fmt.Errorf("error initialising PKCS#12 decoding: %s", C.GoString(C.GetErrorString()))
+	}
+	decoded := C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
```

Comment:
Using `C.ulong(len(data))` instead of `C.size_t(len(data))` gets CI passing on old Go versions. These types of changes are the most frustrating thing about working with Cgo 😡